File: 2020/155.py
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = []
        self.ms = []

    def push(self, x: int) -> None:
        self.stack.append(x)
        if not self.ms or self.ms[-1] >= x:
            self.ms.append(x)

    def pop(self) -> None:
        if self.stack[-1] == self.ms[-1]:
            self.ms.pop()
        self.stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.ms[-1]

# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: current_session/c++/155.cpp
#include <iostream>
#include <stack>

using namespace std;

// simply use 2 vector, to leverage dynamic size array
// push: put everything into snstt, if smaller than last element in mst, put into mst
// pop: pop nst, if equal to mst then pop mst
// top: return nst[-1]
// getMin: return mst[-1]

class MinStack {
 public:
  /** initialize your data structure here. */
  MinStack() {
  }
  
  void push(int val) {
    nStack.push(val);
    if (mStack.empty() || val <= mStack.top()) mStack.push(val);
    return;
  }
  
  void pop() {
    if (!mStack.empty() && nStack.top() == mStack.top()) mStack.pop();
    nStack.pop();
    return;
  }
  
  int top() {
    return nStack.top();
  }
  
  int getMin() {
    return mStack.top();
  }
 private:
  stack<int> mStack;
  stack<int> nStack;
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

int main() {
  MinStack ms = MinStack();
  ms.push(1);
  ms.pop();

}